htdocs/core/lib/notify.lib.php

```diff
@@ -0,0 +1,274 @@
+        complete_substitutions_array($substitutionarray, $outputLangs, $object);
+        $subject = make_substitutions($defaultMessage->topic, $substitutionarray, $outputLangs);
+        $message = make_substitutions($defaultMessage->content, $substitutionarray, $outputLangs);
+    } else {
+        $subject = '['.$companyName.'] '.$outputLangs->transnoentitiesnoconv("DolibarrNotification");
+        $subject .= ($projectTitle ? ' '.$projectTitle : '');
+        $message = $outputLangs->transnoentities("YouReceiveMailBecauseOfNotification", $application, $companyName)."\n";
+        $message .= $outputLangs->transnoentities("YouReceiveMailBecauseOfNotification2", $application, $companyName)."\n";
+        $message .= "\n";
+        $message .= $mesg;
+    }
+
+    $ref = dol_sanitizeFileName($newRef);
+    $pdfPath = $dir_output."/".$ref.".pdf";
+    if (!dol_is_file($pdfPath)||(is_object($defaultMessage) && $defaultMessage->id > 0 && !$defaultMessage->joinfiles)) {
+        // We can't add PDF as it is not generated yet.
+        $pdfFile = '';
+    } else {
+        $pdfFile = $pdfPath;
+        $filename_list[] = $pdfFile;
+        $mimetype_list[] = mime_content_type($pdfFile);
+        $mimefilename_list[] = $ref.".pdf";
+    }
+
+	$parameters = array('notifCode'=>$notifCode, 'to'=>$to, 'from'=>$from, 'file'=>$filename_list, 'mimefile'=>$mimetype_list, 'filename'=>$mimefilename_list, 'outputLangs'=>$outputLangs, 'labelToUse'=>$labelToUse);
+	if (!isset($action)) {
+		$action = '';
+	}
+
+	$reshook = $hookmanager->executeHooks('formatNotificationMessage', $parameters, $object, $action); // Note that $action and $object may have been modified by some hooks
+	if (empty($reshook)) {
+		if (!empty($hookmanager->resArray['subject'])) {
+			$subject .= $hookmanager->resArray['subject'];
+		}
+		if (!empty($hookmanager->resArray['message'])) {
+			$message .= $hookmanager->resArray['message'];
+		}
+	}
+
+	switch ($targetType) {
+		case 'tocontactid':
+			$trackid = 'ctc'.$contactId;
+			$fkContact = 'fk_contact, ';
+			break;
+		case 'touserid':
+			$trackid = 'use'.$contactId;
+			$fkContact = 'fk_user, ';
+			break;
+		default:
+			$trackid = '';
+			$fkContact = '';
+			break;
+	}
+
+	$mailfile = new CMailFile(
+		$subject,
+		$to,
+		$from,
+		$message.' '.$filename_list[0],
+		$filename_list,
+		$mimetype_list,
+		$mimefilename_list,
+		'',
+		'',
+		0,
+		-1,
+		'',
+		'',
+		$trackid,
+		'',
+		'notification'
+	);
+
+	if ($mailfile->sendfile()) {
+		$sql = "INSERT INTO ".$db->prefix()."notify (daten, fk_action, fk_soc, ".$fkContact."type, objet_type, type_target, objet_id, email)";
+		$sql .= " VALUES ('".$db->idate(dol_now())."', ".$actionId.", ".($object->socid > 0 ? ((int) $object->socid) : 'null').", ".(($contactId) ? $contactId.', ' : '')."'".$db->escape($type)."', '".$db->escape($objectType)."', '".$db->escape($targetType)."', ".((int) $object->id).", '".$db->escape($email)."')";
```

Comment:
use (int) before $actionID and $contactID  (all param in sql must use ->escape or (int) or ->idate

Reply:
OK.
Thank you for this feed back!